File: desktop/biwenger_client.py
import requests
import json
import time

class BiwengerClient:
    BASE_URL = "https://biwenger.as.com/api/v2"

    def __init__(self, email, password):
        self.email = email
        self.password = password
        self.session = requests.Session()
        self.token = None
        self.userId = None
        self.leagueId = None
        self.headers = {
            "Content-Type": "application/json",
            "X-Version": "630",
            "X-Client": "android",
            "X-Lang": "es",
            "User-Agent": "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36"
        }

    def login(self):
        try:
            login_url = f"{self.BASE_URL}/auth/login"
            payload = {"email": self.email, "password": self.password}
            response = self.session.post(login_url, json=payload, headers=self.headers)
            if response.status_code == 200:
                data = response.json()
                self.token = data.get("token") or data.get("data", {}).get("token")
                self.headers["Authorization"] = f"Bearer {self.token}"
                
                acc_res = self.session.get(f"{self.BASE_URL}/account", headers=self.headers)
                if acc_res.status_code == 200:
                    acc_data = acc_res.json().get("data", {})
                    self.userId = acc_data.get("id")
                    leagues = acc_data.get("leagues", [])
                    if leagues:
                        self.leagueId = str(leagues[0].get("id"))
                        user_node = leagues[0].get("user")
                        if not self.userId and user_node: self.userId = user_node.get("id")
                    if self.userId: self.headers["X-User"] = str(self.userId)
                    if self.leagueId: self.headers["X-League"] = str(self.leagueId)
                return { "token": self.token, "user_id": self.userId, "league_id": self.leagueId }
            return None
        except:
            return None

    def get_player_list(self):
        try:
            fields = "*,players(*,id),league(*)"
            user_res = self.session.get(f"{self.BASE_URL}/user?fields={fields}", headers=self.headers)
            if user_res.status_code != 200: return [], {}
            
            user_json = user_res.json().get("data", {})
            user_player_ids = {p["id"] for p in user_json.get("players", [])}
            comp_slug = user_json.get("league", {}).get("competition", "la-liga")

            comp_url = f"{self.BASE_URL}/competitions/{comp_slug}/data?score=2"
            comp_res = self.session.get(comp_url, headers=self.headers)
            
            if comp_res.status_code != 200:
                return [], {}

            data = comp_res.json().get("data", {})
            all_players = data.get("players", {})
            all_teams = data.get("teams", {})

            team_id_to_name = {int(tid): t.get("name") for tid, t in all_teams.items()}
            team_id_to_match = {}
            
            for tid, t in all_teams.items():
                tid_int = int(tid)
                next_games = t.get("nextGames", [])
                if next_games:
                    game = next_games[0]
                    h_node = game.get("home")
                    a_node = game.get("away")
                    if h_node and a_node:
                        is_h = int(h_node.get("id")) == tid_int
                        rival_id = int(a_node.get("id") if is_h else h_node.get("id"))
                        team_id_to_match[tid_int] = {
                            "rival": team_id_to_name.get(rival_id, f"Team {rival_id}"),
                            "location": "Home" if is_h else "Away",
                            "date": game.get("date", 0)
                        }

            player_list = []
            for pid_str, p in all_players.items():
                pid = int(pid_str)
                if pid in user_player_ids:
                    fitness = p.get("fitness", [])
                    num_f = [f for f in fitness if isinstance(f, (int, float))]
                    peaks = sum(1 for f in num_f if f >= 9)
                    vol = (max(num_f) - min(num_f)) if len(num_f) > 1 else 0
                    
                    t_id = p.get("teamID") or p.get("team")
                    match_info = team_id_to_match.get(t_id, {"rival": "Unknown", "location": "Unknown", "date": 0})

                    player_list.append({
                        "id": pid,
                        "name": p.get("name"),
                        "position": p.get("position"),
                        "positions": [p.get("position")] + p.get("altPositions", []),
                        "status": p.get("status"),
                        "fitness": fitness,
                        "peaks": peaks,
                        "volatility": vol,
                        "team_id": t_id,
                        "team_name": team_id_to_name.get(t_id, "Unknown"),
                        "rival_name": match_info["rival"],
                        "location": match_info["location"],
                        "match_date": match_info["date"],
                        "price_trend": p.get("priceIncrement", 0)
                    })

            return player_list, team_id_to_name
        except:
            return [], {}

File: mobile/src/logic/BiwengerClient.js
import axios from 'axios';

let lastFetchTime = 0;
let cachedPlayers = null;
let cachedAllPlayers = null;
let cachedTeams = null;

class BiwengerClient {
    constructor(token = null, userId = null, league_id = null) {
        let proxy = "";
        if (typeof window !== 'undefined' && window.location && window.location.hostname === 'localhost') {
            proxy = "https://cors-anywhere.herokuapp.com/";
        }
        this.baseUrl = proxy ? `${proxy}https://biwenger.as.com/api/v2` : "https://biwenger.as.com/api/v2";
        this.token = token;
        this.userId = userId;
        this.leagueId = league_id;
        this.headers = {
            "Content-Type": "application/json",
            "X-Version": "630",
            "X-Client": "android",
            "User-Agent": "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
        };
        if (this.token) this.headers["Authorization"] = `Bearer ${this.token}`;
    }

    async login(email, password) {
        try {
            const res = await axios.post(`${this.baseUrl}/auth/login`, { email, password }, { headers: this.headers });
            const data = res.data;
            this.token = data.token || data.data?.token;
            this.headers["Authorization"] = `Bearer ${this.token}`;
            const accRes = await axios.get(`${this.baseUrl}/account`, { headers: this.headers });
            const accData = accRes.data.data || {};
            let userId = accData.id;
            const leagues = accData.leagues || [];
            if (leagues.length > 0) {
                this.leagueId = String(leagues[0].id);
                if (!userId) userId = leagues[0].user?.id;
            }
            if (userId) this.userId = String(userId);
            return { token: this.token, user_id: this.userId, league_id: this.leagueId };
        } catch (error) { return null; }
    }

    async getPlayerData(forceRefresh = false) {
        const now = Date.now();
        if (!forceRefresh && cachedPlayers && (now - lastFetchTime < 60000)) return cachedPlayers;

        try {
            lastFetchTime = now;
            this.headers["X-User"] = String(this.userId);
            this.headers["X-League"] = String(this.leagueId);
            this.headers["Authorization"] = `Bearer ${this.token}`;

            // ESSENTIAL: Includes 'name' in players(*) fields
            const fields = "*,players(*,fitness,team,owner,status,positions,priceIncrement,name),nextRounds(*),team(*),league(*,points,teams)";
            const res = await axios.get(`${this.baseUrl}/user?fields=${fields}`, { headers: this.headers });
            const userData = res.data.data || {};
            const userPlayerIds = new Set((userData.players || []).map(p => p.id));
            const compSlug = userData.league?.competition || "la-liga";

            const compUrl = `${this.baseUrl}/competitions/${compSlug}/data?score=2`;
            const compRes = await axios.get(compUrl, { headers: this.headers });
            const compData = compRes.data.data || {};
            const allPlayers = compData.players || {};
            const allTeams = compData.teams || {};
            
            let currentRound = 30;
            const activeEvents = compData.activeEvents || [];
            if (activeEvents.length > 0 && activeEvents[0].name) {
                const match = activeEvents[0].name.match(/\d+/);
                if (match) currentRound = parseInt(match[0], 10);
            }
            const remainingMatches = Math.max(0, 38 - currentRound + 1);

            cachedAllPlayers = allPlayers;
            cachedTeams = allTeams;

            const teamNames = {};
            Object.entries(allTeams).forEach(([id, t]) => {
                teamNames[String(id)] = t.name;
            });

            const teamRivals = {};
            Object.entries(allTeams).forEach(([id, t]) => {
                const nextGames = t.nextGames || [];
                if (nextGames.length > 0) {
                    const game = nextGames[0];
                    const h_id = game.home?.id ? String(game.home.id) : null;
                    const a_id = game.away?.id ? String(game.away.id) : null;
                    const current_tid = String(id);
                    if (h_id && a_id) {
                        const isHome = h_id === current_tid;
                        const rivalId = isHome ? a_id : h_id;
                        teamRivals[current_tid] = {
                            rival: teamNames[rivalId] || "Unknown",
                            rivalId: rivalId,
                            location: isHome ? "Home" : "Away"
                        };
                    }
                }
            });

            const playerList = [];
            Object.entries(allPlayers).forEach(([id, p]) => {
                const pid = Number(id);
                // We map all players now so we can use them for the league board
                const tid = String(p.teamID || p.team);
                const matchInfo = teamRivals[tid] || { rival: "Unknown", rivalId: null, location: "Unknown" };
                const fitness = p.fitness || [];
                const num_f = fitness.filter(f => typeof f === 'number');

                const mappedPlayer = {
                    id: pid,
                    name: p.name,
                    position: p.position,
                    positions: [...new Set([p.position, ...(p.altPositions || [])])],
                    status: p.status,
                    fitness: fitness,
                    peaks: num_f.filter(f => f >= 9).length,
                    volatility: num_f.length > 1 ? (Math.max(...num_f) - Math.min(...num_f)) : 0,
                    team_id: tid,
                    team_name: teamNames[tid] || "Unknown",
                    rival_name: matchInfo.rival,
                    rival_id: matchInfo.rivalId,
                    location: matchInfo.location,
                    price_trend: p.priceIncrement || 0,
                    points: p.points || 0,
                    remaining_matches: remainingMatches
                };
                
                // Store in a global cache for league estimations
                allPlayers[id] = mappedPlayer;
                
                if (userPlayerIds.has(pid)) {
                    playerList.push(mappedPlayer);
                }
            });

            cachedPlayers = playerList;
            cachedAllPlayers = allPlayers;
            return playerList;
        } catch (error) {
            if (cachedPlayers) return cachedPlayers;
            throw error;
        }
    }

    async getLeagueBoard() {
        try {
            this.headers["X-User"] = String(this.userId);
            this.headers["X-League"] = String(this.leagueId);
            this.headers["Authorization"] = `Bearer ${this.token}`;
            
            // First ensure we have all players parsed
            if (!cachedAllPlayers) await this.getPlayerData();
            
            // Fetch the competitions data to get active games status
            const compSlug = "la-liga"; // Default
            const compRes = await axios.get(`${this.baseUrl}/competitions/${compSlug}/data?score=5`, { headers: this.headers });
            const activeEvents = compRes.data.data?.activeEvents || [];
            let activeGames = [];
            if (activeEvents.length > 0) {
                activeGames = activeEvents[0].games || [];
            }
            
            const res = await axios.get(`${this.baseUrl}/rounds/league`, { headers: this.headers });
            const data = res.data.data || {};
            
            let activeRound = null;
            if (data.league && data.league.standings) {
                activeRound = data.league;
            } else if (Array.isArray(data)) {
                activeRound = data.find(r => r.status === 'active' || r.status === 'finished') || data[0];
            }
            
            return {
                round: activeRound,
                allPlayers: cachedAllPlayers,
                activeGames: activeGames
            };
        } catch (error) {
            console.error("Failed to get league board:", error);
            return null;
        }
    }
}

export default BiwengerClient;
